executor/src/witgen/data_structures/finalizable_data.rs

```diff
@@ -58,84 +129,185 @@ impl<T: FieldElement> FinalizableData<T> {
-            None => None,
+        if !self.post_finalized_data.is_empty() {
+            self.post_finalized_data.pop()
+        } else if !self.finalized_data.is_empty() {
+            panic!("Row already finalized");
+        } else {
+            self.pre_finalized_data.pop()
+        }
+    }
+
+    /// Removes the last row, even if it has been finalized.
+    pub fn try_remove_last_row(&mut self) {
+        if !self.post_finalized_data.is_empty() {
+            self.post_finalized_data.pop();
+        } else if !self.finalized_data.is_empty() {
+            self.finalized_data.try_remove_last_row();
+        } else {
+            self.pre_finalized_data.pop();
         }
     }
 
     pub fn extend(&mut self, other: Self) {
-        self.data.extend(other.data);
+        if self.finalized_data.is_empty() {
+            self.pre_finalized_data
+                .append(&mut self.post_finalized_data);
+            self.pre_finalized_data.extend(other.pre_finalized_data);
+            self.finalized_data = other.finalized_data;
+            self.post_finalized_data = other.post_finalized_data;
+        } else if other.finalized_data.is_empty() {
+            self.post_finalized_data.extend(other.pre_finalized_data);
+            self.post_finalized_data.extend(other.post_finalized_data);
+        } else if self.post_finalized_data.is_empty() && other.pre_finalized_data.is_empty() {
+            self.finalized_data.data.extend(other.finalized_data.data);
+            self.post_finalized_data = other.post_finalized_data;
+        } else if other.pre_finalized_data.is_empty() {
+            self.finalize_range(
+                (self.pre_finalized_data.len() + self.finalized_data.len())..self.len(),
+            );
+            self.finalized_data.data.extend(other.finalized_data.data);
+            self.post_finalized_data = other.post_finalized_data;
+        } else {
+            panic!(
+                "Cannot extend. Please try to ensure that `other` does not contain finalized rows."
+            );
+        }
     }
 
     pub fn remove(&mut self, i: usize) -> Row<T> {
-        match self.data.remove(i) {
-            Entry::InProgress(row) => row,
-            Entry::Finalized(_) => panic!("Row {i} already finalized."),
+        if i < self.pre_finalized_data.len()
+            && self.finalized_data.is_empty()
+            && self.post_finalized_data.is_empty()
+        {
+            self.pre_finalized_data.remove(i)
+        } else if i < self.pre_finalized_data.len() + self.finalized_data.len() {
+            panic!("Row {i} already finalized.");
+        } else {
+            self.post_finalized_data
+                .remove(i - self.pre_finalized_data.len() - self.finalized_data.len())
         }
     }
 
     pub fn truncate(&mut self, len: usize) {
-        self.data.truncate(len);
+        if len <= self.pre_finalized_data.len() {
+            self.pre_finalized_data.truncate(len);
+            self.finalized_data.clear();
```

Comment:
i don't have the outer context: you can't remove finalized data, but you can truncate it out?

Reply:
I also don't fully understand, but the point is that `pop()` returns a `Row`, which is unfinalized data. So removing the finalized row is not a problem, but we would need to turn it into an unfinalized row again (which is also not a problem per se...)